### ISSUE TITLE
remove `prefer_void_to_null` from recommended

### ISSUE THREAD

**Issue by pq** [dart-lang/lints#154]
In https://github.com/dart-lang/linter/issues/4758, @eernstg proposes we eliminate `prefer_void_to_null` and I'm inclined to agree. Unless there are strong reservations here, I'd propose that as a change to make before publishing 3.0.

/cc @devoncarew @lrhn @natebosch @jakemac53 @goderbauer

**Comment by devoncarew** [comment]
For other readers, the docs for the lint are here:

https://dart.dev/tools/linter-rules/prefer_void_to_null

And in-lined from the dart-lang/linter discussion, from @srawlins:

> I agree I think it has outlived its usefulness; I was getting analysis_server compliant and it was pretty awkward; there were plenty of valid uses of Null.

Which sounds like the lint may be producing more false positives than we'd want.

**Comment by jakemac53** [comment]
It is worth noting that in particular for return types, `void` can be annoying, because it can interact poorly with LUB and other similar computations:

```dart
int y = 1;
var x = switch(y) {
  1 => true,
  _ => print('weird!'); // because this is void, `x` is now inferred as `void`.
};
print(x); // Error, can't use void
```

This is actually a simplification of a real situation that came up recently 🤷

**Comment by goderbauer** [comment]
I am not familiar with all the details here, but I have no objection to removing this lint. While it hasn't caused any trouble in our code base, it also hasn't been all that useful.

**Comment by lrhn** [comment]
I agree with removing this. As discussed elsewhere, there _are_ cases where you should prefer `void` to `Null`, but with null safety, there are more cases where you'd actually choose `Null` deliberately, and you're less likely to choose the wrong one by accident.
Also `null` is no longer so special that it deserves the extra treatment. The same argument applies to every singleton type, whether `()` or a one-element enum (why pass a value to look at, if you know ahead of time what the value will be, it's better to say `void` so that you're not encuraged to actually look at the value.

The problem that this lint solves is now small enough, and well known enough, that it can be handled in code-review instead.

**Comment by natebosch** [comment]
LGTM. I have slight reservations that the same patterns this lint was intended to avoid might start coming back, but I do think it's less useful with null safety, and I do think I recall some places where I wanted to violate this lint.

**Comment by devoncarew** [comment]
This proposal is accepted.
